{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# This is a script to read the sequencing results from the first construct (PGM_1.0 repeat) \n",
    "## Notebook started on the 10th April 2017\n",
    "\n",
    "\n",
    "The construct has 32 repeats that are a series of Emx1.6 targets with 0-1-2 mismatches from the perfect target site (1-16). \n",
    "The same targets are repeated in 17-32 but in this case they have a TAG PAM instead of a TGG.\n",
    "\n",
    "From the first experiment, it was decided that 2 different targets would be selected, a Fast target and a Slow target.\n",
    "The Fast target correspond to the target #16 in the construct, which is the \"perfect target\" (EMx1.6 target).\n",
    "The slow target is the target #1 in the construct.\n",
    "\n",
    "\n",
    "\n",
    "#### The construct was tested in 4 different conditions:\n",
    "-Embryos\n",
    "\n",
    "-Larva (male and Female)\n",
    "\n",
    "-Adult (male and Female)\n",
    "\n",
    "-unmutated target\n",
    "\n",
    "## Sequencing using amplicons\n",
    "\n",
    "The “Amplicon” files contain the results of sequencing by PCR.\n",
    "\n",
    "The repeats were sequenced in triplets, so the first 3 are sequenced in “amplicon 1” and so on..\n",
    "So the Slow target is contained in “Amplicon 1” and the fast target is contained in “Amplicon 6”.\n",
    "\n",
    "#### The sequencing primers for the Fast target are:\n",
    "\n",
    "5' TTGAGAAGATAGACAGAATATGCGC 3'\n",
    "\n",
    "5' TGAGACCGACCCTGACCTGATG 3'\n",
    "\n",
    "\n",
    "\n",
    "## Check the mutation rate for the Untargeted-Fast\n",
    "\n",
    "In here, I will only focus on the sequencing of the Fast target of the Untargeted conditions.\n",
    "\n",
    "### I will manually correct for what we consider a sequencing error:\n",
    "\n",
    "#### THE 3' \"PRIMER\" WILL BE JUST AFTER THE PAM, AND WILL INCLUDE THE POSSIBILITY OF HAVING THE SINGLE NUCLEOTIDE DELETIONS FOUND ON THE UNMUTATED TARGET. These are the 4 posibilities\n",
    "\n",
    "TGGAGTACTGCTGAGACCGAC\n",
    "\n",
    "T-GAGTACTGCTGAGACCGAC\n",
    "\n",
    "TGGAG-ACTGCTGAGACCGAC\n",
    "\n",
    "T-GAG-ACTGCTGAGACCGAC\n",
    "\n",
    "### The read should be at least 100 pb to consider it\n",
    "\n",
    "For the fast target, this is the whole sequenced repeat:\n",
    "\n",
    "( | primer | target | spacer | primer)\n",
    "\n",
    "ttgagaagatagacagaatatgcgc gcgccaccggttgatgtgattgg agtactgc tgagaccgaccctgacctgatg\n",
    "\n",
    "### This is the construct\n",
    "\n",
    "![PGM_1.0_repeats](PGM_1.0_repeats_small.jpg)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.insert(0, \"/usr/lib/python3/dist-packages\")\n",
    "sys.path.insert(0, \"/home/irsalvad/.local/lib/python3.5/site-packages\")\n",
    "#print('\\n'.join(sys.path))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## First, load the packages to use\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import generic_dna\n",
    "import random\n",
    "from collections import Counter\n",
    "random.seed(123)\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import numpy\n",
    "\n",
    "\n",
    "import math\n",
    "from matplotlib import gridspec\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Open the amplicon file\n",
    "\n",
    "As a first step I need to open the file with the results for the \"Untargeted-Fast\". \n",
    "\n",
    "This is in the file ('/Barcode-repeat-split/Untargeted/Untargeted_Amplicon6.txt') provided in the dryad repository (see manuscript for details).\n",
    "\n",
    "Display the first 6 lines of the file\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(line 0) >QP75T:00025:00031\n",
      "\n",
      "(line 1) TTGAGAAGATAGACAGAATATGCGCGCGCCACCGGTTGATGTGATTGAGACTCGCGCGATCG\n",
      "\n",
      "(line 2) >QP75T:00027:00042\n",
      "\n",
      "(line 3) TTGAGAAGATAGACAGAATATGCGCGCGCCACCGGTTGATGTGATTGAGACTGCTCGAGACGA\n",
      "\n",
      "(line 4) >QP75T:00040:00025\n",
      "\n",
      "(line 5) TTGAGAAGATAGACAGAATATGCGCGCGCCACCAATTGATGTGATTAGAGACCTTGCCCGATGCCTGACTAGAAGACCGAA\n",
      "\n"
     ]
    }
   ],
   "source": [
    "input_file ='../Barcode-repeat-split/Untargeted/Untargeted_Amplicon6.txt'\n",
    "Amplicon6 = open(file= input_file)\n",
    "\n",
    "for x in range(0,6):\n",
    "    print(\"(line %d)\" % (x), Amplicon6.readline())\n",
    "Amplicon6.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Upload the primers and define the full construct sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "my primer 5' is TTGAGAAGATAGACAGAATATGCGC\n"
     ]
    }
   ],
   "source": [
    "primer5 = \"TTGAGAAGATAGACAGAATATGCGC\"\n",
    "\n",
    "primer3_1 = \"TGGAGTACTGCTGAGACCGA\"\n",
    "Full_1 = \"TTGAGAAGATAGACAGAATATGCGCGCGCCACCGGTTGATGTGATTGGAGTACTGCTGAGACCGA\"\n",
    "\n",
    "primer3_2 = \"TGAGTACTGCTGAGACCGA\"\n",
    "Full_2 = \"TTGAGAAGATAGACAGAATATGCGCGCGCCACCGGTTGATGTGATTGAGTACTGCTGAGACCGA\"\n",
    "\n",
    "primer3_3 = \"TGGAGACTGCTGAGACCGA\"\n",
    "Full_3 = \"TTGAGAAGATAGACAGAATATGCGCGCGCCACCGGTTGATGTGATTGGAGACTGCTGAGACCGA\"\n",
    "\n",
    "primer3_4 = \"TGAGACTGCTGAGACCGA\"\n",
    "Full_4 = \"TTGAGAAGATAGACAGAATATGCGCGCGCCACCGGTTGATGTGATTGAGACTGCTGAGACCGA\"\n",
    "\n",
    "target = \"TGATGTGAT\"\n",
    "print(\"my primer 5' is\", primer5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define a function to find the primers in each amplified sequence\n",
    "\n",
    "The function takes a line of the file as the input and fisrt check for each base if it matches with the first primer base. If it matches, then looks if the entire primer sequence matches.\n",
    "\n",
    "Then open the file line by line to search for the primer sequence\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## This is the function\n",
    "\n",
    "def find_Primers(dna,my_p5,my_p3):\n",
    "    p5=0; p3=0;\n",
    "    substr = my_p5;\n",
    "    for p in range(len(dna)):\n",
    "        if dna[p] == substr[0]:\n",
    "            if dna[p:p + len(substr)] == substr:\n",
    "                p5=p\n",
    "    substr = my_p3;\n",
    "    for p in range(len(dna)):\n",
    "        if dna[p] == substr[0]:\n",
    "            if dna[p:p + len(substr)] == substr:\n",
    "                p3=p + len(substr)\n",
    "    return p5, p3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Read the file and search line by line for the primers\n",
    "\n",
    "As the file contains the amplified sequence name and then the sequence in the next line read each pair-numbered line (in python would be even-numbered as it starts from 0)\n",
    "\n",
    "Using the function defined previously, search for both primers. The function returns the location of the most 5' and 3' locations of the primers, so the sequence can be spliced.\n",
    "\n",
    "### All sequences found, are appended to the list \"my_seqs\", which will be used for further analysis.\n",
    "### Find the proportion of Unmutated sequences in \"my_seqs\" and store it as a variable (\"No_mut\")\n",
    "\n",
    "\n",
    "\n",
    "(as a test print the fist 5 sequences found)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "line 1 TTGAGAAGATAGACAGAATATGCGCGCGCCACCGGTTGATGTGATTGGAGTACTGCTGAGACCGA\n",
      "\n",
      "Number of unmutated targets = 5417   proportion is 0.9701\n",
      "\n",
      "Total number of targets is  5584\n"
     ]
    }
   ],
   "source": [
    "Amplicon6 = open(file= input_file)\n",
    "\n",
    "my_seqs = list()\n",
    "primer_ends = 0\n",
    "c=0 \n",
    "\n",
    "for i, line in enumerate(Amplicon6):\n",
    "    ## filter out all seqs with length less than 100\n",
    "    if (i % 2 != 0) and (len(line)>=100):\n",
    "        \n",
    "        ## primer 3_1\n",
    "        primer_ends = find_Primers(line,primer5,primer3_1)\n",
    "        if primer_ends != (0,0) :\n",
    "            s = line[primer_ends[0]:primer_ends[1]]\n",
    "            my_seqs.append(s); x = s.find(Full_1)\n",
    "            if x >= 0: c = c+1\n",
    "        ## primer 3_2\n",
    "        primer_ends = find_Primers(line,primer5,primer3_2)\n",
    "        if primer_ends != (0,0) :\n",
    "            s = line[primer_ends[0]:primer_ends[1]]\n",
    "            my_seqs.append(s); x = s.find(Full_2)\n",
    "            if x >= 0: c = c+1\n",
    "        ## primer 3_3\n",
    "        primer_ends = find_Primers(line,primer5,primer3_3)\n",
    "        if primer_ends != (0,0) :\n",
    "            s = line[primer_ends[0]:primer_ends[1]]\n",
    "            my_seqs.append(s); x = s.find(Full_3)\n",
    "            if x >= 0: c = c+1\n",
    "        ## primer 3_4\n",
    "        primer_ends = find_Primers(line,primer5,primer3_4)\n",
    "        if primer_ends != (0,0) :\n",
    "            s = line[primer_ends[0]:primer_ends[1]]\n",
    "            my_seqs.append(s); x = s.find(Full_4)\n",
    "            if x >= 0: c = c+1\n",
    "            \n",
    "                \n",
    "No_mut = c/len(my_seqs)\n",
    "            \n",
    "Amplicon6.close()\n",
    "\n",
    "for i in range(1): print(\"line\",i+1, my_seqs[i])\n",
    "\n",
    "print(\"\\nNumber of unmutated targets =\",c,\"  proportion is %.4g\" % (No_mut))\n",
    "print(\"\\nTotal number of targets is \", len(my_seqs))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#out1 = open(\"my_seqs.txt\",\"w\") \n",
    "#for index, item in enumerate(my_seqs): \n",
    "#    out1.write(my_seqs[index])\n",
    "#    out1.write(\"\\n\")\n",
    "#out1.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Define a function to search for the position of the primer\n",
    "\n",
    "#### The output of the function is the first nucleotide of the seq primer. This will be used to read the Nmers\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def find_target(dna,myprimer3):\n",
    "    sub = None\n",
    "    substr = myprimer3\n",
    "    for p in range(len(dna)):\n",
    "        if dna[p] == substr[0]:\n",
    "            if dna[p:p + len(substr)] == substr:\n",
    "                sub=p\n",
    "    return sub"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get all possible sequences from the sequencing primer if you read 9bp from the 3' primer\n",
    "\n",
    "This is to know how many different deletions we have, and what is their distribution (e.g., is there one copy that is found in 25% of the sequences?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of Nmers: 18 total: 5584 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "Nmer_pb = 9 ## Number of nucleotides in the target\n",
    "N_unique = 0;\n",
    "c= 0;\n",
    "Nmer_LIST=list();\n",
    "## Search in all amplified sequences:\n",
    "for s in range(len(my_seqs)):\n",
    "    seq = my_seqs[s]\n",
    "    # primer 3_1\n",
    "    x = find_target(seq,primer3_1)\n",
    "    if x != None:             # if the primer can be found in the amplified sequence\n",
    "        c = c + 1; Nmer_LIST.append(seq[x-Nmer_pb:x])\n",
    "    # primer 3_2\n",
    "    x = find_target(seq,primer3_2)\n",
    "    if x != None:             # if the primer can be found in the amplified sequence\n",
    "        c = c + 1; Nmer_LIST.append(seq[x-Nmer_pb:x])\n",
    "    # primer 3_3\n",
    "    x = find_target(seq,primer3_3)\n",
    "    if x != None:             # if the primer can be found in the amplified sequence\n",
    "        c = c + 1; Nmer_LIST.append(seq[x-Nmer_pb:x])\n",
    "    # primer 3_4\n",
    "    x = find_target(seq,primer3_4)\n",
    "    if x != None:             # if the primer can be found in the amplified sequence\n",
    "        c = c + 1; Nmer_LIST.append(seq[x-Nmer_pb:x])\n",
    "\n",
    "UniqueNmers = set(Nmer_LIST)\n",
    "N_unique = len(UniqueNmers)\n",
    "\n",
    "Nmer_TOTALS = c;\n",
    "print(\"number of Nmers:\", len(UniqueNmers), \"total:\", Nmer_TOTALS,\"\\n\") \n",
    "#print(c, N_unique, UniqueNmers)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get the frequency of each of the Nmers for all the sites\n",
    "### Use the unmutated proportion of targets as a reference. The ideal would be to have the lowest possible frequency of the \"unmutated\" Nmer, and that get many mutated Nmers at a frequency that is informative"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 3, 3, 6, 13, 22, 5522]]\n"
     ]
    }
   ],
   "source": [
    "Sorted_Lists_Vals = list()\n",
    "Sorted_Lists_Labs = list()\n",
    "Original_Nmers = list()\n",
    "\n",
    "Unique_Nmer_freq= list()\n",
    "\n",
    "### Get the real Nmer\n",
    "Nmer_seq = target\n",
    "\n",
    "UniqueList = list(UniqueNmers)\n",
    "#print(\"for \", \"distance:\", len(UniqueList), \"\\n\") \n",
    "\n",
    "## The Nmer_List contains all Nmers, search the frequency of each unique Nmer (uniques are in UniqueList)\n",
    "Nmer_freq=list()\n",
    "for i in range(len(UniqueList)):\n",
    "    Unique_Nmer_freq.append([UniqueList[i],\n",
    "                  #           '{:04d}'.format(Nmer_LIST.count(UniqueList[i]))]\n",
    "                             Nmer_LIST.count(UniqueList[i])])\n",
    "    #print(Unique_Nmer_freq)\n",
    "Sorted_Unique = sorted(Unique_Nmer_freq, key=lambda Unique_Nmer_freq: Unique_Nmer_freq[1])\n",
    "Sorted_Unique_labs = [item[0] for item in Sorted_Unique]\n",
    "Sorted_Unique_vals = [item[1] for item in Sorted_Unique]\n",
    "\n",
    "Sorted_Lists_Vals.append(Sorted_Unique_vals)\n",
    "Sorted_Lists_Labs.append(Sorted_Unique_labs)\n",
    "   \n",
    "print(Sorted_Lists_Vals)\n",
    "#print(Sorted_Lists_Labs)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot all the Nmer freqs \n",
    "\n",
    "In the plot below, the frequency (Yaxis) of each Nmer (Xaxis) is shown. The Nmers are sorted from low to high frequency to see ir more clearly. In red we can see the proportion of Nmers with a sequence equal to the \"unmutated\" Nmer (Could be the unmutated or not).\n",
    "As a dashed vertical line, we can see the real proportion of unmutated sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Freq of the most common Nmer: 0.9889 \t the unmutated total: 5522 proportion : 0.9889\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAY4AAAEZCAYAAACAZ8KHAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAH3VJREFUeJzt3XmUXGW57/HvLxMCAg0i4AkhQdGg3EDjEBBZ0BiB4EU5\nB66coNdDIwdxIOqRw6DXY0B0HdHlgCOoSMApiF4VPAoo0MeLGAhgQchgGGOYkjDJsCA2yXP/2Luz\ni6a6u6q6qnZV7d9nrV5d795vVz31rEo92e9Te5ciAjMzs2pNyDsAMzPrLC4cZmZWExcOMzOriQuH\nmZnVxIXDzMxq4sJhZmY1ceEwM7OauHBYYUi6T9JaSVuWbTtR0nU5xrRJ0lOSnkx/P5ZXLGbVcuGw\nIgmS1/zHKmxvKkkTR9gVwN4RsW1EbBMRO9T492Yt58JhRfNF4FRJ21bamR4BfFDSKkl/k/QZSa+U\n9EdJT0haJGlS2fwjJf1Z0uOSrpc0q2zfvZJOl3Qb8LSkSv/elP4Mj+NgSWvSv38I+H4Vj7evpFvS\nuBdJ+omkz9SdKbMRuHBY0dwMDACnjTLnMGBfYH/gdOAC4N3ANGAWcBwkb9TAhcBJwA7pvMslTS67\nr3nAEUBPRGyqMdZdgB5gN+D9oz1e+pi/AC5O910GHFPj45lVxYXDimgBcIqkl42w/9yIeCYiVgB3\nAFdHxOqIeAr4LUlRgeQN/PyIuDkSPwA2kBScIedFxIMRsWGUeG5NjyAek/TVsu0bgQURMZj+/WiP\ntz8wKSK+FhEbI+LnwJKasmJWpUljTzHrLhGxTNKvgU8AKypMWVd2+1lg7bDxzunt6cC/SJqfjgVM\nBv6hbP79VYS0b0TcW2H7+ogYLBuP9XgPDPv71VU8tlnNfMRhRXUWyf/gp47jPtYAn4uIHdKf7SPi\npRFxadmcahrvL+pxjPC3oz3eQ7z4uexW1bMwq5ELhxVSRNwNXAp8ZBx3813gA5JmA0jaWtLbJW3d\niBhrfLw/Ac9Lmi9pkqSjgdlNisMKzoXDimT4/+A/A2w1bPvwOSMeMUTELSRHLd9Iz79YBRxfzd/W\nOGfMx0uXtI4GTgAeBd4F/Lza+zarhfL+IidJFwJHAmsjYu8K+98NnJEOnwI+GBFLWxiiWUeSdBGw\nJiI+nXcs1l3a4YjjIuDwUfbfAxwUEfsAnyU5XDczs5zk/qmqiLhe0vRR9i8uGy5mfM1MsyLx90Jb\nU+ReOGr0rySfozezMUTE+/KOwbpTxxQOSYeQNP4OzDsWM7Mi64jCIWlv4DvA3Ih4fIQ5Piw3M6tD\nRIx0LlFF7dAchxEu9AYgaTeSjxW+N/3s/Ygiwj8RLFiwIPcY2uXHuXAunIvRf+qR+xGHpB8DfcDL\nJP2V5DpCU4CIiO8A/0Fy0bZvSRIwGBE+sWkU9913X94htA3nIuNcZJyL8cm9cETEu8fYfxLJSU9m\nZtYG2mWpyhqov78/7xDahnORcS4yzsX45H7meKNIim55LmZmrSKJ6NDmuDXQwMBA3iG0Deci41xk\nnIvxceEwM7OaeKnKzKzAvFRlZmZN58LRhbx+m3EuMs5FxrkYHxcOMzOriXscZmYF5h6HmZk1nQtH\nF/L6bca5yDgXGedifFw4zMysJu5xmJkVmHscZmbWdC4cXcjrtxnnIuNcZJyL8XHhMDOzmrjHYWZW\nYO5xmJlZ07lwdCGv32aci4xzkXEuxseFw8zMauIeh5lZgbnHYWZmTefC0YW8fptxLjLORca5GJ/c\nC4ekCyWtlXT7KHO+JulOSSVJva2Mz8zMXij3HoekA4GngUsiYu8K+48ATomI/ylpP+C8iNi/wjz3\nOMzMatSRPY6IuB54fJQpRwGXpHNvBLaTtHMrYjMzsxfLvXBUYSqwpmz8QLrNRuD124xzkXEuMq3I\nxZo1azh1/nwOmz2bU+fPZ82aNWP/UYeYlHcAjdTf38+MGTMA6Onpobe3l76+PiB7oXhcrPGQdokn\nz3GpVGqrePIcl0qlpt7/T3/6Uz504omcsGEDHx8c5KJbb2XfhQv58/LlTJs2LdfnPzAwwMKFCwE2\nv1/WKvceB4Ck6cAVI/Q4zgeui4hL0/FK4OCIWDtsnnscZtYWTp0/nwkXXMBHB3fiOV7CHtzNaZMn\ns+nkk/nS17+ed3gv0JE9jpTSn0ouB/4FQNL+wBPDi4aZWTtZeuONzBkcZBHz+DYfBGDO4CBLb7op\n58gaI/fCIenHwA3AayT9VdIJkk6W9H6AiPgNcK+ku4ALgA/lGG5HGL5MU2TORca5yDQ7F7P2249r\nJk9mHTuxE+sAuGbyZGbNnt3Ux22V3HscEfHuKuacUs19SS8+aBlp+arSXM/3fM/3/EbO30s7c2is\n5LTJk/nxS1/K4tNPzzWeRi3nt0WPoxEkVXwi7fZC8nzP9/zizN99+h1sN+VC3nr4IB87/XSmTZuW\nazyV5tfT4+iqwtEtz8XMusOb3gTf/Ca08wpVJzfHrYG8lp1xLjLORaZVuVi3DnbaqSUP1VIuHGZm\nTRCRFI6XvzzvSBrPS1VmZk3w1FOwyy7wzDN5RzI6L1WZmbWJ9eu7c5kKXDi6kteyM85FxrnItCIX\n3drfABcOM7Om6ObC4R6HmVkTfO978Kc/wYUX5h3J6NzjMDNrE918xOHC0YW8lp1xLjLORaZVPY5u\n/CguuHCYmTVFN3+qyj0OM7MmOPRQOO00OOywvCMZnXscZmZtwj0O6yhey844FxnnIuPzOMbHhcPM\nrME2bYJHHoEdd8w7kuZwj8PMrMEeewxe9Sp4/PG8IxmbexxmZm2gm5epwIWjK3ktO+NcZJyLTLNz\n4cJhZmY16fbC4R6HmVmDfetbsHQpfPvbeUcyNvc4zMzaQLcfcbhwdCGvZWeci4xzkWl2Ltav797r\nVEEbFA5JcyWtlLRK0hkV9k+TdK2kWyWVJB2RR5xmZtXq9iOOXHsckiYAq4A5wIPAEmBeRKwsm3MB\ncGtEXCDptcBvImL3CvflHoeZtYWDD4azz4a+vrwjGVsn9jhmA3dGxOqIGAQWAUcNm7MJ2Da93QM8\n0ML4zMxq1u1HHHkXjqnAmrLx/em2cmcD75W0Bvg1ML9FsXUsr2VnnIuMc5HxeRzjMynvAKpwHHBR\nRHxF0v7AD4G9Kk3s7+9nxowZAPT09NDb20tfeqw49ELxuFjjIe0ST57jUqnUVvHkOS6VSk27/8FB\neOKJAW6/Hd761vZ4vuXjgYEBFi5cCLD5/bJWefc49gfOioi56fhMICLi3LI5dwCHR8QD6fhuYL+I\neGTYfbnHYWa5e/hh2GcfWLs270iq04k9jiXAHpKmS5oCzAMuHzZnNfA2gLQ5vsXwomFm1i66fZkK\nci4cEbEROAW4GlgGLIqIFZLOlnRkOu3fgZMklYAfAcfnE23nGL5MU2TORca5yDQzF0UoHLn3OCLi\nSmDmsG0Lym6vAA5sdVxmZvUoQuHwtarMzBroq1+Fe++F887LO5LqdGKPw8ysqxThiMOFowt5LTvj\nXGSci0wzc9Ht16kCFw4zs4YqwhGHexxmZg305jfDl74EBxyQdyTVcY/DzCxnRTjicOHoQl7LzjgX\nGeci4/M4xseFw8ysQZ59FgYHYZtt8o6kudzjMDNrkL/+Fd7yFlizZuy57cI9DjOzHBVhmQpcOLqS\n17IzzkXGucg0KxcuHGZmVpOiFA73OMzMGuQLX0jOHP/iF/OOpHrucZiZ5agoRxwuHF3Ia9kZ5yLj\nXGSalYsiXKcKXDjMzBqmKEcc7nGYmTXIG94AF1wAb3xj3pFUzz0OM7McFeWIw4WjC3ktO+NcZJyL\nTDNyEZEUDvc4zMysKk8+CVtsAVtumXckzeceh5lZA9x1Fxx+ONx9d96R1MY9DjOznBSlvwEuHF3J\na9kZ5yLjXGSakQsXjhaSNFfSSkmrJJ0xwpxjJS2TtFTSD1sdo5nZWIpUOHLtcUiaAKwC5gAPAkuA\neRGxsmzOHsClwCER8aSkHSPikQr35R6HmeXms59Nvsjpc5/LO5LadGKPYzZwZ0SsjohBYBFw1LA5\nJwHfjIgnASoVDTOzvBXpiCPvwjEVKP+urPvTbeVeA8yUdL2kGyQd3rLoOpTXsjPORca5yDQjF0W5\nThXApLwDqMIkYA/gIGA34A+S/sfQEUi5/v5+ZsyYAUBPTw+9vb309fUB2QvF42KNh7RLPHmOS6VS\nW8WT57hUKjX8/v/yFzjxxPZ4fqONBwYGWLhwIcDm98ta5d3j2B84KyLmpuMzgYiIc8vmfBtYHBEX\np+PfA2dExC3D7ss9DjPLzaxZ8KMfwd575x1JbTqxx7EE2EPSdElTgHnA5cPm/BI4BEDSjsCrgXta\nGqWZ2Rjc42iRiNgInAJcDSwDFkXECklnSzoynXMV8KikZcA1wL9HxOO5Bd0Bhi/TFJlzkXEuMo3O\nxaZN8NhjsOOODb3btpV7jyMirgRmDtu2YNj4VODUVsZlZlatxx6D7baDSbm/o7aGr1VlZjZOy5fD\nMcfAihV5R1K7TuxxmJl1vCL1N8CFoyt5LTvjXGSci0yjc+HCYWZmNSla4RizxyHp6NH2R8T/bWhE\ndXKPw8zy8ulPw8SJsGDB2HPbTT09jmo+A3AicABwbTo+BLgBWA8E0BaFw8wsL+vXd96Jf+NRzVLV\nZOB1EXFMRBwD7AVMjogTIuJ9zQ3P6uG17IxzkXEuMs3ocRTlOlVQXeGYFhEPlY3XklwzyszMcI/j\nxROkb5Bc5uMn6aZ/Bu6KiPlNjq0m7nGYWV5mzoRf/Qr23DPvSGpXT4+jqhMAJf0TydVpAf4QEb+o\nI76mcuEws7xsvz3cfTfssEPekdSumScA3gr8V0T8G3CVpG1qjs5axmvZGeci41xkGpmLv/8dnn4a\nenoadpdtb8zCIekk4GfABemmqSRXrDUzK7xHHkka4xMKdFZcNT2OEslXvN4YEfum25ZGxKwWxFc1\nL1WZWR5KJTj+eLjttrwjqU+zlqo2RMTfyx5kEsn5G2ZmhVe0T1RBdYXjvyV9EthS0qHAZcAVzQ3L\nxsNr2RnnIuNcZBqZCxeOys4kOUt8KXAy8BvgU80MysysUxSxcIza45A0EbgkIt7TupDq4x6HmeXh\nzDOTL3H6xCfyjqQ+De9xpF/tOvR94GZmNsz69cW63AhUt1R1D/BHSf8h6eNDP80OzOrnteyMc5Fx\nLjLucYzPiIVD0g/Sm+8Efp3O3absx8ys8IpYOEbscUhaDrwNuBLoG74/Ih5ramQ1co/DzPKw++5w\nzTXwylfmHUl9Gv19HOcD1wC7AzeXPw7JeRwdmiYzs8Yp4hHHiEtVEfG1iHgtcFFEvLLsZ/eIcNFo\nY17LzjgXGeci06hcPPMMRMDWWzfk7jrGmM3xiPhgMwOQNFfSSkmrJJ0xyrxjJG2S9PpmxmNmVq2h\nT1SppoWezlfVZdWb9uDSBGAVMAd4EFgCzIuIlcPmvRT4L5JvIzwlIm6tcF/ucZhZS910E3z4w7Bk\nSd6R1K+Zl1VvltnAnRGxOiIGgUXAURXmnQN8HtjQyuDMzEZTxP4G5F84pgJrysb3p9s2k7QvsGtE\n/LaVgXUyr2VnnIuMc5FpVC6KWjhG+1RV7iQJ+DJwfPnmkeb39/czY8YMAHp6eujt7aWvrw/IXige\nF2s8pF3iyXNcKpXaKp48x6VSqSH3t25dHzvtlP/zqWU8MDDAwoULATa/X9Yq7x7H/sBZETE3HZ8J\nREScm463Be4CniYpGLsAjwLvHN7ncI/DzFrt4x+HqVPh1FPzjqR+ndjjWALsIWnoeljzgMuHdkbE\nkxGx09DHgIHFwDsqNcfNzFqtiNepgpwLR3oRxVOAq4FlwKKIWCHpbElHVvoTRlmqssTwZZoicy4y\nzkWmUblwjyMnEXElMHPYtgUjzH1rS4IyM6tCUQtHrj2ORnKPw8xabepUuPFG2HXXvCOpXz09DhcO\nM7M6RMAWW8BTTyW/O1UnNsetCbyWnXEuMs5FphG5eOIJ2Gqrzi4a9XLhMDOrQ1E/UQVeqjIzq8v1\n18MZZ8Af/5h3JOPjpSozsxYp6ieqwIWjK3ktO+NcZJyLTCNy4cJhZmY1KXLhcI/DzKwOp5wCM2fC\n/Pl5RzI+7nGYmbVIkT9V5cLRhbyWnXEuMs5Fxj2O8XHhMDOrQ5ELh3scZmZ1ePnLYdmyzi8evlZV\nlzwXM2tvzz8PW24Jzz0HEyfmHc34uDlugNeyyzkXGeciM95cPPoobL995xeNerlwmJnVqMifqAIv\nVZmZ1ezaa+Gcc+C66/KOZPy8VGVm1gJF/kQVuHB0Ja9lZ5yLjHORGW8uXDjMzKwmRS8c7nGYmdXo\n5JPh9a9Pfnc69zjMzFpg3bpif6rKhaMLeS0741xknIuMexzjk3vhkDRX0kpJqySdUWH/v0laJqkk\n6XeSpuURp5nZkKIXjlx7HJImAKuAOcCDwBJgXkSsLJtzMHBjRDwn6QNAX0TMq3Bf7nGYWUtstx2s\nXg09PXlHMn6d2OOYDdwZEasjYhBYBBxVPiEi/jsinkuHi4GpLY7RzGyzDRvg2WeT4lFUeReOqcCa\nsvH9jF4YTgR+29SIuoDXsjPORca5yIwnF+vXJ8tUqun/6N1lUt4BVEvS/wbeABw80pz+/n5mzJgB\nQE9PD729vfT19QHZC8XjYo2HtEs8eY5LpVJbxZPnuFQq1f3369bBS14ywMBA+zyfWsYDAwMsXLgQ\nYPP7Za3y7nHsD5wVEXPT8ZlARMS5w+a9DTgPOCgiHh3hvtzjMLOmu/JK+MpX4Kqr8o6kMTqxx7EE\n2EPSdElTgHnA5eUTJO0LnA+8c6SiYWbWKkX/RBXkXDgiYiNwCnA1sAxYFBErJJ0t6ch02heArYHL\nJP1Z0i9zCrdjDF+mKTLnIuNcZMaTCxeONuhxRMSVwMxh2xaU3T605UGZmY3AhcPXqjIzq8kJJ8BB\nByW/u0En9jjMzDpK0a9TBS4cXclr2RnnIuNcZNzjGB8XDjOzGrhwuMdhZla1CNhqK3jkEdh667yj\naQz3OMzMmujpp2HixO4pGvVy4ehCXsvOOBcZ5yJTby6GrlNVdC4cZmZV8ieqEu5xmJlV6fLL4bvf\nhSuuyDuSxnGPw8ysifyJqoQLRxfyWnbGucg4F5l6c+HCkXDhMDOrkgtHwj0OM7Mqvec98Pa3J7+7\nhXscZmZN5E9VJVw4upDXsjPORca5yLjHMT4uHGZmVXLhSLjHYWZWhU2bYIst4JlnYMqUvKNpHPc4\nzMya5PHHYZttuqto1MuFowt5LTvjXGSci0w9ufB1qjIuHGZmVfAnqjLucZiZVeFnP4Of/AR+/vO8\nI2ks9zjMzJrEn6jKuHB0Ia9lZ5yLjHORqScXLhyZ3AuHpLmSVkpaJemMCvunSFok6U5Jf5K0Wx5x\nmlmxuXBkcu1xSJoArALmAA8CS4B5EbGybM4HgVkR8SFJ/wz8U0TMq3Bf7nGYWdO8611w7LHJ725S\nT49jUrOCqdJs4M6IWA0gaRFwFLCybM5RwIL09s+Ab7Q0QjPrWhEwODj6z/PPJ7/vvdefqhqSd+GY\nCqwpG99PUkwqzomIjZKekLRDRDzWohg7yt/+BscdN8ArXtGXdyht4aGH2jsXQwfJES+8Pdq2eq1d\nO8DOO/fVHV8tv+sx9Hw3bUp+Kt2utG3jxuTN/fnnX3h7tG0bNgywaVMfGzfCpEkwefILfypt23JL\nmDmz/ufXTfIuHPUY8ZCqv7+fGTNmANDT00Nvby99fX1A1gzr9vF++/Wx114AyXjPPZP9K1cWc7zD\nDrDnnu0TT6WxlIwleO1rX7j/da9L9q9Yke2XYPnybD9UN5ZKHHFE9fPLHz/5+xeOAfbaKxkvW5bs\nLx8P7Yfqx7Nm9TFhAtxxxwATJsA++yT3d8cdyf59903233Zbsn/fffuYNCkZT5yYvP4nTYJbb03G\nBxyQjG+6KRkffHAfEyfC+eeXeP3rYc6c5P6r/fc19J+Qdvn3Xs94YGCAhQsXAmx+v6xV3j2O/YGz\nImJuOj4TiIg4t2zOb9M5N0qaCDwUES9qUbnHYWZWu048j2MJsIek6ZKmAPOAy4fNuQI4Pr39LuDa\nFsZnZmbD5Fo4ImIjcApwNbAMWBQRKySdLenIdNqFwI6S7gQ+BpyZT7Sdw5/XzzgXGeci41yMT+49\njoi4Epg5bNuCstsbgGNbHZeZmVXma1WZmRVYJ/Y4zMysw7hwdCGv32aci4xzkXEuxseFw8zMauIe\nh5lZgbnHYWZmTefC0YW8fptxLjLORca5GB8XDjMzq4l7HGZmBeYeh5mZNZ0LRxfy+m3Gucg4Fxnn\nYnxcOMzMrCbucZiZFZh7HGZm1nQuHF3I67cZ5yLjXGSci/Fx4TAzs5q4x2FmVmDucZiZWdO5cHQh\nr99mnIuMc5FxLsbHhcPMzGriHoeZWYG5x2FmZk2XW+GQtL2kqyX9RdJVkrarMGcfSTdIWiqpJOnY\nPGLtNF6/zTgXGeci41yMT55HHGcCv4+ImcC1wCcqzHkGeG9EzAKOAL4qadsWxtiRSqVS3iG0Deci\n41xknIvxybNwHAVcnN6+GPjH4RMi4q6IuDu9/RCwDnh5yyLsUE888UTeIbQN5yLjXGSci/HJs3Ds\nFBFrASLiYWCn0SZLmg1MHiokZmaWj0nNvHNJvwN2Lt8EBPCpCtNH/EiUpFcAlwDvbWiAXeq+++7L\nO4S24VxknIuMczE+uX0cV9IKoC8i1kraBbguIl5bYd42wADw2Yj4xSj358/impnVodaP4zb1iGMM\nlwP9wLnA8cCvhk+QNBn4JXDxaEUDan/iZmZWnzyPOHYAfgpMA1YDx0bEE5LeAJwcEe+X9B7g+8Ay\nsmWu/oi4PZegzcyse84cNzOz1uiqM8clLZB0v6Rb05+5ecfUapLmSlopaZWkM/KOJ0+S7pN0m6Q/\nS7op73haSdKFktZKur1s25gn3XajEXJRyPcKSbtKulbSsvTE6o+k22t6bXTVEYekBcBTEfHlvGPJ\ng6QJwCpgDvAgsASYFxErcw0sJ5LuAd4QEY/nHUurSToQeBq4JCL2TredCzwaEV9I/1OxfUScmWec\nrTBCLgr5XpF+EGmXiChJeilwC8k5dSdQw2ujq444UkVuks8G7oyI1RExCCwieVEUlejO1/iYIuJ6\nYHjBHPOk2240Qi6ggO8VEfFwRJTS208DK4BdqfG10Y3/qD6cXtfqe0U5FC8zFVhTNr4/3VZUAVwl\naYmkk/IOpg3UdNJtART5vQJJM4BeYDGwcy2vjY4rHJJ+J+n2sp+l6e93AN8CXhURvcDDQKEOQ+1F\n3hIRbwTeTvImcWDeAbWZ7lmnrl2h3yvSZaqfAR9NjzyGvxZGfW3keR5HXSLi0Cqnfhe4opmxtKEH\ngN3Kxrum2wopvb4ZEbFe0i9IlvKuzzeqXK2VtHPZSbfr8g4oLxGxvmxYqPcKSZNIisYPImLo/Lma\nXhsdd8QxmvQJDzkauCOvWHKyBNhD0nRJU4B5JCdaFo6krdL/VSFpa+Awivd6EC9cxx866RZGOOm2\ni70gFwV/r/g+sDwizivbVtNro9s+VXUJyZrdJuA+khMJ1+YaVIulHys8j+Q/BRdGxOdzDikXknYH\nfkFyyD0J+FGRciHpx0Af8DJgLbCA5CoMlzHspNu8YmyVEXJxCAV8r5D0FuAPwFKSfxsBfBK4iQon\nZI94P91UOMzMrPm6aqnKzMyaz4XDzMxq4sJhZmY1ceEwM7OauHCYmVlNXDjMzKwmLhzW8SRtkvTF\nsvGpkj6dZ0zVkvRUFXNqOttd0sGSCnMmtLWeC4d1gw3A0em3SjZNetn6kfZNrPNuxzyRKiLqucaW\nT9CypnHhsG7wPPAd4OPDd0i6SNK3JP1J0l3p/8YvlLRc0vfL5h0q6QZJN0u6VNJW6fZ7JX1e0s3A\n/6pw39+WtBg4V9Kb0vu4RdL1kl6dzjte0s8l/Tb9opxzK8S5Y/q3R1TY91T6+2BJ10m6TNIKST8o\nmzM33XYzySU0hrZvlT7fxWlc70i3f0zShentWenFQl9SU9atsDruIodmFQTwTWBppTdloCci3izp\nnSTX5HlzRCxPi8TeJBeC/BQwJyKelXQ6SRH6bPr3j6RX2a1kakTsD5uvOHpgRGySNAf4T7Jisw/J\nJS4Ggb9I+lpEPJD+3U5pXJ+MiGtHeH5DeoHXkVzR9Y+SDiD5Mp7vAH0RcY+kS8vm/x/gmog4Mb10\n+E2Sfk9yWZrrJP1jOuekiHhuhOdo9gIuHNYVIuJpSRcDHwWeHbZ7aL1/KfBwRCxPx8uAGSTX53kd\nyRuxgMnADWV/X/5GPNxlZbd7gEvSI42ha2QNuSa9fDWSlgPTSQrWFOD3wIcj4v9V8VRvGrrqr6RS\nGv8zwD0RcU8654fA0PePHAa8Q9Jp6XgKsFtE/EXSCcDtwPkRsbiKxzYDXDisu5wH3ApcNGz7hvT3\nprLbQ+NJ6e+rI+I9I9zvM6M8Zvm+c4BrI+JoSdOB6yrEALCR7N/e8yRHDHOBagrHSPcz0rfZCTgm\nIu6ssO81wFPAP1TxuGabucdh3UAA6XeL/xQ4cay5wywG3iLpVbC5L/DqOuLYluz7T06o8m8CeB+w\nZ7pEVslYX3G6EpieXhEY4LiyfVcBH9l8R1Jv+ns7kkJ7EPAyScdUGa+ZC4d1hfIewJdILp8dFfYN\nHwdARDxC8l0EP5F0G8ky1cwR/n6k+wL4IvB5Sbcw+r+tF8QQySWqjwMOkfSBKh5nePwbgJOB36TN\n8fLLg58DTE6/JfMO4DPp9i8DX4+Iu4B/Bf5T0o6jxGy2mS+rbmZmNfERh5mZ1cSFw8zMauLCYWZm\nNXHhMDOzmrhwmJlZTVw4zMysJi4cZmZWExcOMzOryf8HwqciBzf7MwcAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7fb5a54c0940>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "Vals = Sorted_Lists_Vals[0]\n",
    "    \n",
    "Labs = Sorted_Lists_Labs[0]\n",
    "    \n",
    "Rel_Vals = list()\n",
    "for index, item in enumerate(Vals):\n",
    "    Rel_Vals.append(Vals[index]/Nmer_TOTALS)\n",
    "        \n",
    "plt.plot(Rel_Vals)\n",
    "plt.xlabel('Nmer rank index')\n",
    "plt.ylabel('freq')\n",
    "plt.title('Nmer Freq')\n",
    "plt.grid(True)\n",
    "#plt.axvline(x=Nmer90, ymin=0, ymax=1,linewidth=1, c='black', hold=None,linestyle=\"dashed\")\n",
    "\n",
    "plt.axhline(y=No_mut, xmin=0, xmax=1,linewidth=2, c='black', hold=None,linestyle=\"dashed\")\n",
    "## plot the sequence that is the same as the unmutated one\n",
    "for index, item in enumerate(Labs):\n",
    "    if item == target:\n",
    "        #ax.annotate(Nmer,(index,Vals[index]), size = 14,color = \"r\")\n",
    "        plt.scatter(index,Rel_Vals[index],s=30,c=\"r\")\n",
    "        if Rel_Vals[index] < 1 :\n",
    "            print(\"Freq of the most common Nmer: %.5g\" % (max(Vals)/Nmer_TOTALS),\n",
    "                  \"\\t the unmutated total:\", Vals[index],\"proportion : %.5g\" % (Vals[index]/Nmer_TOTALS) )\n",
    "\n",
    "ax = plt.gca()\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Important for the simulations:\n",
    "\n",
    "## Probability of a seq Error (After N cell cycles):\n",
    "\n",
    "#### APROXIMATION USING NMERS    (1- unmutated Nmer )       = 1 - 0.9889 = 0.0111\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
